river/neighbors/knn_classifier.py

```diff
@@ -1,176 +1,193 @@
+    @property
+    def _multiclass(self):
+        return True
 
-    def learn_one(self, x, y):
-        """Update the model with a set of features `x` and a label `y`.
+    def clean_up_classes(self) -> "KNNClassifier":
+        """
+        Classes that are added (and removed) from the window may no longer be valid.
+        This method cleans up the window and and ensures only known classes
+        are added, and we do not consider "None" a class. It is called every
+        `cleanup_every` step, or can be called manually.
+        """
+        self.classes = {x for x in self.window if x[0][1] is not None}
+
+    def learn_one(self, x: dict, y=None):
+        """Learn a set of features `x` and optional class `y`.
+        Parameters:
+            x: A dictionary of features.
+            y: A class (optional if known).
+        Returns:
+            self
+        """
+        # Only add the class y to known classes if we actually add the point!
+        if self.nn.update((x, y), n_neighbors=self.n_neighbors):
+            self.classes.add(y)
 
-        Parameters
-        ----------
-        x
-            A dictionary of features.
-        y
-            The class label.
+        # Ensure classes known to instance reflect window
+        self._run_class_cleanup()
+        return self
 
-        Returns
-        -------
-            self
+    def _run_class_cleanup(self):
+        """
+        Helper function to run class cleanup, accounting for _cleanup_counter.
+        """
+        # clean up classes every cleanup_every steps
+        if self.cleanup_every:
+            self._cleanup_counter -= 1
+            if self._cleanup_counter == 0:
+                self.clean_up_classes()
+                self._cleanup_counter = self.cleanup_every
 
-        Notes
-        -----
-        For the K-Nearest Neighbors Classifier, fitting the model is the
-        equivalent of inserting the newer samples in the observed window,
-        and if the size_limit is reached, removing older results.
+        return self
 
+    def predict_one(self, x: dict):
+        """Predict the label of a set of features `x`.
+        Parameters:
+            x: A dictionary of features.
+        Returns:
+            The neighbors
         """
+        return self.predict_proba_one(x)
 
-        self.classes_.add(y)
-        x_arr = dict2numpy(x)
+    def predict_proba_one(self, x):
+        """Predict the class of a set of features `x`.
+        Parameters:
+            x: A dictionary of features.
+        Returns:
+            Lookup (dict) of classes and probability predictions (normalized)
+        """
+        nearest = self.nn.find_nearest((x, None), n_neighbors=self.n_neighbors)
 
-        self.data_window.append(x_arr, y)
+        # Default prediction for every class we know is 0.
+        # If class_cleanup is false this can include classes not in window
+        y_pred = {c: 0.0 for c in self.classes}
```

Comment:
I don't think this is correct. A probability distribution should sum to 1. I would instead default to `1 / n_classes`.

Reply:
This is verbatim from creme, so that is wrong too? https://github.com/MaxHalford/creme/blob/0ceffdee9e5afbb22de57f3666417015c71334be/creme/neighbors/knn.py#L158-L160